### ISSUE TITLE
Support for keyset pagination in DataProvider

### ISSUE THREAD

**Issue by heruan** [vaadin/flow#6196]
Currently the DataProvider uses the concept of skip/limit to query the backend, but this could become highly inefficient for large data when the backend is a SQL server since how skip (or offset) works. This and an efficient solution is described in [1], which can be resolved with a better pagination strategy known as "keyset pagination": asking the backend for a certain number of items after a specific item.

To implement this, the DataProvider would provide the same limit value already present in the query, but also the item (or its ID from `DataProvider#getId`) from which to start.

[1] http://allyouneedisbackend.com/blog/2017/09/24/the-sql-i-love-part-1-scanning-large-table/

**Comment by heruan** [comment]
I'd like to contribute to this, but I need some guidance. Keyset pagination is all about sorting: given a set of sort orders and a limit size, you need the sort values of the "pivot" element (usually the last element of the previous page), e.g.

SQL backend:
```sql
select * from users order by name, id limit 10
select * from users where (name, id) > ('foo', 33) order by name, id limit 10
...
select * from users where (name, id) > ('zed', 876) order by name, id limit 10
```

MongoDB backend:
```js
users.sort({'name': 1, '_id': 1}).limit(10)
users.find({'name': {'$gt': 'foo'}, '_id': {'$gt': 33}}).sort({'name': 1, '_id': 1}).limit(10)
...
users.find({'name': {'$gt': 'zed'}, '_id': {'$gt': 876}}).sort({'name': 1, '_id': 1}).limit(10)
```

To enable a `DataProvider` to do so, the `Query` should provide the last item of the current set to be able to get its sorting values to the query, something like:

```java
query -> {
	var pivot = query.getPivot(); // <- new method
	var limit = query.getLimit();
	var order = query.getSortOrders();
	return backend.fetchWithPivot(pivot, limit, order);
}
```

I'm looking at `DataCommunicator` and I see the query is created in `fetchFromProvider`, but I'm unsure how to get the pivot item from there: I guess it should be retrievable from the `keyMapper`, but how?

**Comment by heruan** [comment]
A less opinionated approach could be simpler: just let to set a generic `keyset` object when fetching, later to be used for subsequent queries; e.g.

```java
query -> {
	var keyset = query.getKeyset(); // can be null, e.g. first query
	var limit = query.getLimit();
    var skip = query.getSkip(); // fallback if keyset isn't available
	var order = query.getSortOrders();
	var page = backend.fetchWithKeyset(keyset, limit, skip, order);
}
```

**Comment by Springrbua** [comment]
I don't think that this feature can simply be added to the existing DataProvider, as it also affects the UI.
In the `vaadin-grid` for example, you have a Scrollbar and as you scroll, the offset will be calculated, using the total amount of items and the scroll position. So you know the `offset`, but not the `keyset` for the query.
I have created a [similar issue](https://github.com/vaadin/vaadin-grid-flow/issues/675) a while ago, with the addition, that the total count is also unknown.

In my opinion, the `keyset` pagination should not be part of the current `DataProvider` but they should be separate. As I said before, `keyset` pagination will also have an impact on the UI / UX, as an exact movement to a given position (offset) won't be possible.

**Comment by mperktold** [comment]
I'd like to add that keyset pagination doesn't just provide better performance, but also "better correctness" in case the dataset changes between fetching one page and another.

For example, suppose you are looking at items 51 to 100 while someone inserts 3 items between 50 and 51. Now your range effectively becomes 54 to 103, but the Grid doesn't know that. So when you navigate back and fetch the previous page, you get 1 to 50 instead of 4 to 53, skipping over the newly inserted rows.

With keyset pagination, you don't have this problem.

One special case of this problem is when you are looking at the first page, say 1 to 50, and you want to go to the "previous" page because you think there might be some new entries. With offset pagination, this would result in a negative offset which of course dosn't make sense. With keyset pagination there is no problem, you can always ask what's before the item you see first.

The remaining question is if and how the UI lets the user do this. Currently, the only way is to refresh all data.
